### PR TITLE
ICU-21480 integrate CLDR release-39-alpha0 to ICU trunk for ICU 69rc

### PR DESCRIPTION
**Pull request by pedberg-icu**
##### Checklist

- [x] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21480
- [x] Updated PR title and link in previous line to include Issue number
- [x] Issue accepted
- [x] Tests included
- [x] Documentation is changed or added

This integrates the CLDR release-39-alpha0 to ICU trunk for ICU 69rc. There are several changes that are either significant or affect a lot of files; in some cases I have noted people that I think should look at them in more detail.
- The CLDR tools build process has changed to use maven, which affects portions of the CLDR-to-ICU integration and ICU data build process that use or refer to CLDR tools. The changes here are only a partial step towards what eventually needs to be done; @srl295 has done some further work on this and should look at the changes here.
- [CLDR-2698](https://unicode-org.atlassian.net/browse/CLDR-2698) effectively swapped the behavior of locales for _no_ and _nb_ to match the way other macrolanguages are handled (like zh and cmn) so that all the content formerly in _nb_ has been moved to _no_, and the _nb_ is now just a stub that aliases to _no_. However, there is an additional wrinkle: _no_ is also treated as the parent language for _nn_, so that things like formats and collation can be inherited by _nn_ from _no_. Also, we retain a localeDisplayName for _nb_ that is distinct from the name for _no_. This may have significant implications for some clients. @FrankYFTang  should look at this (perhaps especially at the changes to localeCanonicalization.txt) ; @jefgen should also look at this, and particularly the change to Windows LCID mapping in icu4c/source/common/locmap.cpp. On MSVC this leads to the following ICU4C cintltst failure:
    ```
	cldrtest  {
	 TestLocaleStructure   {
	  (Known issue 20181) ICU-20181 Fix LCID mapping for ckb vs ku
	   ERROR:   no_NO has 0x0414 and the number resolves wrongfully to nb_NO
	 } ---[1 ERRORS in /tsutil/cldrtest/TestLocaleStructure]  (140ms)  
	 ...
    ```
- [CLDR-14355](https://unicode-org.atlassian.net/browse/CLDR-14355) and [CLDR-13566](https://unicode-org.atlassian.net/browse/CLDR-13566) made changes to locale matcher behavior; in particular the former eliminated fallbacks between Hans and Hant, which affected a number of tests. @markusicu should check these.
- Some other CLDR changes affected many files:
    - The symbol for currency XOF was changed in all locales, affecting curr data
    - The name for "root" was removed in all locales, affecting lang data
    - The CLDR unit type "concentr-milligram-per-deciliter" was changed to "concentr-milligram-ofglucose-per-deciliter", affecting units data (and "concentr-ofglucose" was added). May need to have a discussion about how to handle the API names for creating MeasureUnits for this.
- There were some changes related to how unit preferences for consumption were handled.
    - On the CLDR side, [PR-991](https://github.com/unicode-org/cldr/pull/991) eliminated the category "consumption-inverse" and folded it into "consumption", and [PR-1010](https://github.com/unicode-org/cldr/pull/1010) updated unitPreferencesTest.txt accordingly.
    - Meanwhile, on the ICU side, [PR-1550](https://github.com/unicode-org/icu/pull/1550) made some ICU changes to implement inverse handling for consumption. These two sets of changes had a conflict which caused a test failure, but that is now resolved by rebasing onto latest master which includes merged [PR-1565](https://github.com/unicode-org/icu/pull/1565). Thanks @hugovdm!

### CODE REVIEW

**Comment by srl295** [comment]
can we switch to the @hagbard tool instead?

**Comment by srl295** [comment]
@hagbard @pedberg-icu here is my change for maven: (this obviates the `install-cldr-jars.sh`  script) https://github.com/srl295/icu/commit/b6b7a26b5a71d97c67fe04cd6fcdfa92476b29f6

**Comment by pedberg-icu** [comment]
How would this impact icu4c/source/data/build.xml <https://github.com/unicode-org/icu/pull/1551#diff-8ee3400d326317a2b6f2decc6b3d50da881a569a9dddd67cad1817b8ca33a9b9> which was looking directly in $CLDR_DIR/tools/java/ for classes/ and cldr.jar, and is now (in my version) looking in  $CLDR_DIR/tools/cldr-code/target/ for classes/ and cldr-code.jar?

- Peter E

**Comment by srl295** [comment]
@pedberg-icu please see https://github.com/pedberg-icu/icu/pull/1

**Comment by srl295** [comment]
> How would this impact icu4c/source/data/build.xml <[#1551 (comment)](https://github.com/unicode-org/icu/pull/1551#diff-8ee3400d326317a2b6f2decc6b3d50da881a569a9dddd67cad1817b8ca33a9b9)> which was looking directly in $CLDR_DIR/tools/java/ for classes/ and cldr.jar, and is now (in my version) looking in $CLDR_DIR/tools/cldr-code/target/ for classes/ and cldr-code.jar? - Peter E
> […](#)

as far as i can tell that build.xml is obsolete and i've removed it in my updated PR-against-your-PR

**Comment by srl295** [comment]
@hagbard VERY STRANGE.

I kept getting 'cannot write to data file'  as below

```
     [exec] BUILD FAILED
     [exec] /Users/srl295/src/icu/tools/cldr/cldr-to-icu/build-icu-data.xml:142: java.lang.RuntimeException: cannot write ICU data file: ug
     [exec]     at org.unicode.icu.tool.cldrtoicu.IcuTextWriter.writeToFile(IcuTextWriter.java:59)
     [exec]     at org.unicode.icu.tool.cldrtoicu.LdmlConverter.write(LdmlConverter.java:600)
     [exec]     at org.unicode.icu.tool.cldrtoicu.LdmlConverter.processLdml(LdmlConverter.java:354)
     [exec]     at org.unicode.icu.tool.cldrtoicu.LdmlConverter.convertAll(LdmlConverter.java:237)
     [exec]     at org.unicode.icu.tool.cldrtoicu.LdmlConverter.convert(LdmlConverter.java:195)
     [exec]     at org.unicode.icu.tool.cldrtoicu.ant.ConvertIcuDataTask.execute(ConvertIcuDataTask.java:342)
     [exec]     at org.apache.tools.ant.UnknownElement.execute(UnknownElement.java:292)
     [exec]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
     [exec]     at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:64)
     [exec]     at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
     [exec]     at java.base/java.lang.reflect.Method.invoke(Method.java:564)
     [exec]     at org.apache.tools.ant.dispatch.DispatchUtils.execute(DispatchUtils.java:99)
     [exec]     at org.apache.tools.ant.Task.perform(Task.java:350)
     [exec]     at org.apache.tools.ant.Target.execute(Target.java:449)
     [exec]     at org.apache.tools.ant.Target.performTasks(Target.java:470)
     [exec]     at org.apache.tools.ant.Project.executeSortedTargets(Project.java:1401)
     [exec]     at org.apache.tools.ant.Project.executeTarget(Project.java:1374)
     [exec]     at org.apache.tools.ant.helper.DefaultExecutor.executeTargets(DefaultExecutor.java:41)
     [exec]     at org.apache.tools.ant.Project.executeTargets(Project.java:1264)
     [exec]     at org.apache.tools.ant.Main.runBuild(Main.java:827)
     [exec]     at org.apache.tools.ant.Main.startAnt(Main.java:223)
     [exec]     at org.apache.tools.ant.launch.Launcher.run(Launcher.java:284)
     [exec]     at org.apache.tools.ant.launch.Launcher.main(Launcher.java:101)
     [exec] Caused by: java.nio.file.FileAlreadyExistsException: /Users/srl295/src/icu/tools/cldr/cldr-to-icu/../../../icu4c/source/data/curr/ug.txt
     [exec]     at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:94)
     [exec]     at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:106)
     [exec]     at java.base/sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:111)
     [exec]     at java.base/sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:218)
     [exec]     at java.base/java.nio.file.spi.FileSystemProvider.newOutputStream(FileSystemProvider.java:478)
     [exec]     at java.base/java.nio.file.Files.newOutputStream(Files.java:224)
     [exec]     at java.base/java.nio.file.Files.newBufferedWriter(Files.java:2997)
     [exec]     at org.unicode.icu.tool.cldrtoicu.IcuTextWriter.writeToFile(IcuTextWriter.java:54)
     [exec]     ... 22 more
     [exec] 
```

So i solv… er, i didn't solve anything, but i hacked around it below:


```diff
diff --git a/tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/IcuTextWriter.java b/tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/IcuTextWriter.java
index 93f23524c9..c2d61904be 100644
--- a/tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/IcuTextWriter.java
+++ b/tools/cldr/cldr-to-icu/src/main/java/org/unicode/icu/tool/cldrtoicu/IcuTextWriter.java
@@ -56,7 +56,13 @@ static void writeToFile(
                 new IcuTextWriter(icuData).writeTo(out, header);
             }
         } catch (IOException e) {
-            throw new RuntimeException("cannot write ICU data file: " + icuData.getName(), e);
+            if(e instanceof java.nio.file.FileAlreadyExistsException && allowOverwrite == false) {
+                System.err.println("re-overwriting file " + icuData.getName() + "  -concurrency?");
+                writeToFile(icuData, outDir, header, true);
+                return;
+            } else {
+                throw new RuntimeException("cannot write ICU data file: " + icuData.getName(), e);
+            }
         }
     }
 ```

… But putting this code in, caused the issue to not get hit. where it was consistently happening before.

Not sure.

**Comment by srl295** [comment]
With my PR above and the patches (`nb… ` and also the overwrite) i am able to complete a generation against cldr 39

**Review by srl295 — COMMENTED** [review]
LG so far, but see my subordinate PR

**Comment by pedberg-icu** [comment]
Thanks for the subordinate PR, I will take a look at that after I get further with this (mainly fixing test failures, the integration works but I have not pushed all of the relevant files yet).

**Comment by srl295** [comment]
I think the goal right now is just to get you unstuck, so I would say do whatever works and then we'll improve it￼

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsConverter.java
Comment (hugovdm) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsConverter.java:124]: Thanks for adding these!

**Review by hugovdm — APPROVED** [review]
Units-related changes all look good to me. (Including CLDR-sourced data _content_ changes for languages I know.)

**Review by hugovdm — COMMENTED** [review]
-> Busy looking at the units unit test failures.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsConverter.java
Comment (srl295) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitsConverter.java:124]: more details @ CLDR-14232

**Review by srl295 — DISMISSED** [review]
- build files and test code LGTM
- data RSLGTM
Looks like only 'golden master' type test changes, which is great.

**Review by hugovdm — COMMENTED** [review]
Should fix the ICU4J fuel consumption conversions: https://github.com/unicode-org/icu/pull/1565

**Comment by jefgen** [comment]
Re: no and nb 
> This may have significant implications for some clients. @jefgen should also look at this, and particularly the change to Windows LCID mapping in icu4c/source/common/locmap.cpp

Thanks for flagging this @pedberg-icu.

Right now, Windows currently does not have any specific locale(s) for the language code `no`, so there isn't a `no-NO` locale. This is likely why the `TestLocaleStructure` test is failing in the CI builds.

From the logs:

```
ERROR:   no_NO has 0x0414 and the number resolves wrongfully to nb_NO
 ---[1 ERRORS in /tsutil/cldrtest/TestLocaleStructure]  (116ms)  
```

However, I don't yet know what the plans are for how Windows will handle the nb/no change.

If the locale is renamed and keeps the same LCID -- with an alias added for nb -- then the issue for the ICU test code is that it would only be on future versions, so older versions of Windows there would still be the locale `nb_NO`. 
So the test would need to handle *both* cases for the LCID 0x0414, where it either round-trips or does not round-trip.

On the other hand, if the locale is not renamed, and it stays as `nb-NO` without any aliases (ie: no change), then the locale name won't round-trip ever.

That said, since the Windows locale for LCID 0x0414 hasn't yet been changed and/or renamed at all, I'm wondering if the locale in `locmap.cpp` should really be changed at the moment (?).

I'm thinking that this might need to be to be handled with another `log_knownIssue` to skip the test for `no` for now.

(Note: There is already an existing one for the ku vs ckb locale here: https://github.com/unicode-org/icu/blob/0583afebf33cffe4a8560f64acfa75b3e14f34d8/icu4c/source/test/cintltst/cldrtest.c#L488 )

I'm going to file a separate ticket ( [ICU-21488](https://unicode-org.atlassian.net/browse/ICU-21488) ), as I think we'll likely need one for other potential future changes (based on what direction Windows takes), which can also be used for a `log_knownIssue` for now.

Also, cc'ing @ShawnSteele and @daniel-ju as well, in case I'm missing something, or if they have more comments as well.

**Comment by pedberg-icu** [comment]
Rebased onto latest master which includes merged [PR-1565](https://github.com/unicode-org/icu/pull/1565), that should fix the test failures.

File: icu4c/source/data/lang/nn.txt
Comment (markusicu) [inline icu4c/source/data/lang/nn.txt:5]: Shouldn't some of the "nn" strings be redundant with the new "no" parent and be removed from the "nn" files?

**Review by markusicu — DISMISSED** [review]
lgtm tnx!

File: icu4c/source/data/lang/nn.txt
Comment (pedberg-icu) [reply icu4c/source/data/lang/nn.txt:5]: I think that will happen when we do the data minimization as part of CLDR BRS, that has not yet happened

**Comment by hugovdm** [comment]
FYI I'm causing conflicts, having changed some ldml rules and data files: I've merged #1527, and would like to also merge #1566 - and might as well also merge it, after which I think you'll need to rerun the build-icu-data.xml rules.

Apologies for this coming in so late, causing a bit of extra work for this integration!

**Comment by hugovdm** [comment]
(If github is up to date, it looks like I didn't cause any conflicts. Might also be reasonable to just submit this PR as is then! After which I can also check whether I've caused any diffs in build-icu-data's output - hopefully not.)

**Comment by pedberg-icu** [comment]
We will be revising the approach to handling no/nb in a separate PR, but we should go ahead and merge this to unblock other work.

**Comment by jefgen** [comment]
> We will be revising the approach to handling no/nb in a separate PR, but we should go ahead and merge this to unblock other work.

Thank you @pedberg!

FYI @erik0686: I think once this PR is merged the version number update PR could be rebased onto master, and then the *.jar files could be re-generated. :)

**Review by hugovdm — APPROVED** [review]

**Comment by pedberg-icu** [comment]
For re-reviewers (people who already approved an earlier version), the only recent changes here are:
- rebase onto current master of of early on Feb 10
- in main/classes/core/src/com/ibm/icu/impl/number/LongNameHandler.java PluralTableSink, skip handling of "case" and "gender" keys (flagged with "// TODO: @Hugo to fix for new grammatical stuff")
- add a logKnownIssue to skip the Windows LCID mapping failure for no_NO in  cldrtest/TestLocaleStructure

**Review by jefgen — APPROVED** [review]
thanks peter! -- lgtm with the logKnownIssue for nb/no for now.

**Pull request merged**
